crates/iceberg/src/transform/mod.rs

```diff
@@ -0,0 +1,51 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Transform function used to compute partition values.
+use crate::{spec::Transform, Result};
+use arrow::array::ArrayRef;
+
+mod identity;
+mod temporal;
+mod void;
+
+/// TransformFunction is a trait that defines the interface for all transform functions.
+pub trait TransformFunction: Send {
+    /// transform will take an input array and transform it into a new array.
+    /// The implementation of this function will need to check and downcast the input to specific
+    /// type.
+    fn transform(&self, input: ArrayRef) -> Result<ArrayRef>;
```

Comment:
I'm not a rust'er, but is this passing in an array?

Maybe some context on how this is implemented in Python (and the Java implementation also followed this path eventually). In PyIceberg we initialize a transform, but that one might be unbound. A common place to find a transform is in the partition spec: https://iceberg.apache.org/spec/#partition-specs
Here we have a reference to a field by the `source-id`. This means that we don't know yet what kind the input is. For example, you can bucket a string, uuid, long, int, datetime, etc. Once we bind the transform to the schema, we return the actual function as a callable (we pass a function): 

```python
    def transform(self, source: IcebergType, bucket: bool = True) -> Callable[[Optional[Any]], Optional[int]]:
        source_type = type(source)
        if source_type in {IntegerType, LongType, DateType, TimeType, TimestampType, TimestamptzType}:

            def hash_func(v: Any) -> int:
                return mmh3.hash(struct.pack("<q", v))

        elif source_type == DecimalType:

            def hash_func(v: Any) -> int:
                return mmh3.hash(decimal_to_bytes(v))

        elif source_type in {StringType, FixedType, BinaryType}:

            def hash_func(v: Any) -> int:
                return mmh3.hash(v)

        elif source_type == UUIDType:

            def hash_func(v: Any) -> int:
                if isinstance(v, UUID):
                    return mmh3.hash(v.bytes)
                return mmh3.hash(v)

        else:
            raise ValueError(f"Unknown type {source}")

        if bucket:
            return lambda v: (hash_func(v) & IntegerType.max) % self._num_buckets if v else None
        return hash_func
```
This way we can initialize the transform, even when we don't know the type of the source type.

Reply:
There are two differences here:
1. The transform function is the actual implementation of transform logic, it's similar to `hash_func` above in python.
2. The bind process happens in following `create_transform_function`. The difference is that in rust we don't accept a `Type` parameter, since it's included in `ArrayRef`.
3. We implement the partition function in arrow array rather iceberg value. This is because currently the most important use case for us it to implement `PartitionedWriter`, which accepts arrow `RecordBatch`. And implement it on arrow array let use do it in vectorized approach(much faster than row based approach). I guess eventuall we will need to implement an other version in iceberg value and it would be useful in partition pruning.

Reply:
Currently this transform logic is specific for arrow record batch. Maybe we need to consider a interface more generic? 

Actually I think the partition writer is used by external system, and arrow is a good middle layer between iceberg and external system. If we want to consider iceberg value, we need to take a effor to maintain convert of iceberg value. Is there benefit for using iceberg value?🤔

Reply:
For example when we want to execute `select * from t where a = 1 and b=2`, and table `t` is partitioned by a, we can prune unnecessary partitions by calculating `a=1`.

Reply:
Sorry, I think my question should be whether we need to provide write interface use iceberg value. We should maintain iceberg value internally for patition and convert between arrow.

Reply:
For writer, I don't think we should provide it.

Reply:
> When a file is being written, Iceberg keeps the upper and lower bounds of each of the columns. By default, a truncate(16) transform is applied before storing the metrics in the manifest.

Did you mean [the lower bound(upper bound) in field summary](https://iceberg.apache.org/spec/#manifests:~:text=NaN%20value%20for%20the%20field-,optional,field%2C%20or%20null%20if%20all%20values%20are%20null%20or%20NaN%20%5B2%5D,-optional)? So it's the lower bound of original column and then apply transform to it `transform(lower_bound(col))` rather than lower bound of transform column `lower_bound(transform(col))`.

Reply:
Hi, @Fokko For more background on why we use an array can be found here: https://github.com/apache/iceberg-rust/issues/32#issuecomment-1697308710.

But I do think transform functions accepting single value would also be useful, maybe we can implement it in following pr?

Reply:
Yes. And it's better to integrate transform array and single value, I can do it when implementation of single value version.